compiler/generator_php.c

```diff
@@ -0,0 +1,1406 @@
+            continue;
+        }
+        ch = input[i++];
+        switch (ch) {
+            case '~': write_char(g, '~'); continue;
+            case 'f': write_block_start(g);
+                      write_failure(g);
+                      g->unreachable = false;
+                      write_block_end(g);
+                      continue;
+            case 'M': write_margin(g); continue;
+            case 'N': write_newline(g); continue;
+            case '{': write_block_start(g); continue;
+            case '}': write_block_end(g); continue;
+            case 'S': {
+                int j = input[i++] - '0';
+                if (j < 0 || j > (int)(sizeof(g->S) / sizeof(g->S[0]))) {
+                    printf("Invalid escape sequence ~%c%c in writef(g, \"%s\", p)\n",
+                           ch, input[i - 1], input);
+                    exit(1);
+                }
+                write_string(g, g->S[j]);
+                continue;
+            }
+            case 'B': {
+                int j = input[i++] - '0';
+                if (j < 0 || j > (int)(sizeof(g->B) / sizeof(g->B[0])))
+                    goto invalid_escape2;
+                write_s(g, g->B[j]);
+                continue;
+            }
+            case 'I': {
+                int j = input[i++] - '0';
+                if (j < 0 || j > (int)(sizeof(g->I) / sizeof(g->I[0])))
+                    goto invalid_escape2;
+                write_int(g, g->I[j]);
+                continue;
+            }
+            case 'V':
+                write_varref(g, p->name);
+                continue;
+            case 'W':
+                write_varname(g, p->name);
+                continue;
+            case 'L':
+                write_literal_string(g, p->literalstring);
+                continue;
+            case '+': g->margin++; continue;
+            case '-': g->margin--; continue;
+            case 'n': write_string(g, g->options->name); continue;
+            case 'P': write_string(g, g->options->parent_class_name); continue;
+            case 'C': { // Constant.
+                w(g, "/* TODO */ const"); // ??
+                continue;
+            }
+            case 'D': { // Declare variable.
+                w(g, "/* TODO ?*/" ); // ??
+                continue;
+            }
+            default:
+                printf("Invalid escape sequence ~%c in writef(g, \"%s\", p)\n",
+                       ch, input);
+                exit(1);
+            invalid_escape2:
+                printf("Invalid escape sequence ~%c%c in writef(g, \"%s\", p)\n",
+                       ch, input[i - 1], input);
+                exit(1);
+        }
+    }
+}
+
+static void w(struct generator * g, const char * s) {
+    writef(g, s, NULL);
+}
+
+static void generate_AE(struct generator * g, struct node * p) {
+    const char * s;
+    switch (p->type) {
+        case c_name:
+            write_string(g,"$this->");
+            write_varref(g, p->name); break;
```

Comment:
The `$this->` should be output by `write_varref()` rather than by everything that calls it - that's the difference between `write_varref()` (a reference to a variable) and `write_varname()` (just the name of the variable, e.g. for when we declare it).  For some languages the two functions (correctly) generate the same thing.

Reply:
Fixed in b92d555